sockeye/data_io.py

```diff
@@ -1477,10 +1546,109 @@ def load_state(self, fname: str):
+
+    def iter_next(self) -> bool:
+        """
+        True if the iterator can return another batch.
+        """
+
+        # Read batch_size lines from the source stream
+        sources_sentences = [[] for x in self.sources_sentences]
+        target_sentences = []
+        num_read = 0
+        for num_read, (sources, target) in enumerate(parallel_iter(self.sources_sentences, self.target_sentences, reiterate=False), 1):
+            source_len = len(sources[0])
+            target_len = len(target)
+            if source_len > self.max_len_source:
+                logger.info("Trimming source sentence {} ({} -> {})".format(self.sentno + num_read, source_len, self.max_len_source))
+                sources = [source[0:self.max_len_source] for source in sources]
+            if target_len > self.max_len_target:
+                logger.info("Trimming target sentence {} ({} -> {})".format(self.sentno + num_read, target_len, self.max_len_target))
+                target = target[0:self.max_len_target]
+
+            for i, source in enumerate(sources):
+                sources_sentences[i].append(source)
+            target_sentences.append(target)
+            if num_read == self.batch_size:
+                break
+
+        self.sentno += num_read
+
+        if num_read == 0:
+            self.next_batch = None
+            return False
+
+        dataset = self.data_loader.load(sources_sentences,
+                                        target_sentences,
+                                        [num_read]).fill_up(self.bucket_batch_sizes, self.fill_up)
+
+        data = [dataset.source[0], dataset.target[0]]
+        label = dataset.label
+
+        provide_data = [mx.io.DataDesc(name=n, shape=x.shape, layout=C.BATCH_MAJOR) for n, x in
+                        zip(self.data_names, data)]
+        provide_label = [mx.io.DataDesc(name=n, shape=x.shape, layout=C.BATCH_MAJOR) for n, x in
+                         zip(self.label_names, label)]
+
+        self.next_batch = mx.io.DataBatch(data, label,
+                                          pad=0, index=None, bucket_key=self.buckets[0],
+                                          provide_data=provide_data, provide_label=provide_label)
+
+        return True
+
+    def next(self) -> mx.io.DataBatch:
+        """
+        Returns the next batch.
+        """
+        if self.iter_next():
+            return self.next_batch
+        raise StopIteration
+
+    def save_state(self, fname: str):
+        raise Exception('Not supported!')
+
+    def load_state(self, fname: str):
+        raise Exception('Not supported!')
+
+
 class ShardedParallelSampleIter(BaseParallelSampleIter):
     """
     Goes through the data one shard at a time. The memory consumption is limited by the memory consumption of the
-    largest shard. The order in which shards are traversed is changed with each reset.
+    largest shard. The order in whcirh shards are traversed is changed with each reset.
```

Comment:
is this a vim command gone wrong? :)

Reply:
When in doubt, blaming vim is a good idea.